pf/internal/schemav6/schemav6.go

```diff
@@ -0,0 +1,231 @@
+	res, ok := resp.ResourceSchemas[resourceName]
+	contract.Assertf(ok, "schemav6.ResourceSchema: no %q resource in GetProviderSchema response", resourceName)
+
+	return res, nil
+}
+
+func DataSourceSchema(schema dschema.Schema) (*tfprotov6.Schema, error) {
+	psrv := providerserver.NewProtocol6(&schemaExtractingProvider{
+		dsSchemas: map[string]dschema.Schema{
+			dsName: schema,
+		}})
+
+	resp, err := psrv().GetProviderSchema(context.Background(), &tfprotov6.GetProviderSchemaRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := renderDiagnostics(resp.Diagnostics); err != nil {
+		return nil, err
+	}
+
+	res, ok := resp.DataSourceSchemas[dsName]
+	contract.Assertf(ok, "schemav6.DataSourceSchema: no %q datasource in GetProviderSchema response", dsName)
+
+	return res, nil
+}
+
+type schemaExtractingProvider struct {
+	providerSchema  *pschema.Schema
+	resourceSchemas map[string]rschema.Schema
+	dsSchemas       map[string]dschema.Schema
+}
+
+func (*schemaExtractingProvider) Metadata(context.Context, provider.MetadataRequest, *provider.MetadataResponse) {
+}
+
+func (p *schemaExtractingProvider) Schema(_ context.Context, _ provider.SchemaRequest, resp *provider.SchemaResponse) {
+	if p.providerSchema != nil {
+		resp.Schema = *p.providerSchema
+	}
+}
+
+func (*schemaExtractingProvider) Configure(context.Context, provider.ConfigureRequest, *provider.ConfigureResponse) {
+}
+
+func (p *schemaExtractingProvider) DataSources(context.Context) []func() datasource.DataSource {
+	keys := mapKeys(p.dsSchemas)
+	ret := []func() datasource.DataSource{}
+	for _, k := range keys {
+		k := k
+		ret = append(ret, func() datasource.DataSource {
+			return &schemaExtractingDS{p.dsSchemas[k]}
+		})
+	}
+	return ret
+}
+
+func (p *schemaExtractingProvider) Resources(context.Context) []func() resource.Resource {
+	keys := mapKeys(p.resourceSchemas)
+	ret := []func() resource.Resource{}
+	for _, k := range keys {
+		k := k
+		ret = append(ret, func() resource.Resource {
+			return &schemaExtractingResource{p.resourceSchemas[k]}
+		})
+	}
+	return ret
+}
+
+var _ provider.Provider = (*schemaExtractingProvider)(nil)
+
+type schemaExtractingResource struct {
+	rschema rschema.Schema
+}
+
+func (r *schemaExtractingResource) Metadata(
+	_ context.Context,
+	_ resource.MetadataRequest,
+	resp *resource.MetadataResponse,
+) {
+	resp.TypeName = resourceName
+}
+
+func (r *schemaExtractingResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = r.rschema
+}
+
+func (r *schemaExtractingResource) Create(context.Context, resource.CreateRequest, *resource.CreateResponse) {
+}
+
+func (r *schemaExtractingResource) Read(context.Context, resource.ReadRequest, *resource.ReadResponse) {
+}
+
+func (r *schemaExtractingResource) Update(context.Context, resource.UpdateRequest, *resource.UpdateResponse) {
+}
+
+func (r *schemaExtractingResource) Delete(context.Context, resource.DeleteRequest, *resource.DeleteResponse) {
+}
```

Comment:
Can we assert that these methods will not be called (and then panic)?

Reply:
I wonder if this would reduce branch coverage :)